### ISSUE TITLE
KokkosBlas::dot is broken for complex, due to incorrect assumptions about Fortran ABI

### ISSUE THREAD

**Issue by mhoemmen** [kokkos/kokkos-kernels#307]
@kyungjoo-kim @crtrott See comment here: https://github.com/trilinos/Trilinos/issues/3493#issuecomment-426044331

**Comment by mhoemmen** [comment]
See this article on the relevant functions: http://mirror.informatimago.com/next/developer.apple.com/hardware/ve/errata.html

**Comment by mhoemmen** [comment]
https://www.math.utah.edu/software/c-with-fortran.html#function-return-types

"...you should not expect to use Fortran functions that return types such as COMPLEX or COMPLEX*16. Write a SUBROUTINE interface to your Fortran function instead, and then invoke it as a void function from C or C++."

**Comment by mhoemmen** [comment]
My Trilinos PR https://github.com/trilinos/Trilinos/pull/3538 only minimally patches kokkos-kernels to fix this issue.  I leave it to the kokkos-kernels developers to figure out the best longer-term approach.  However, please do fix this before the next develop-to-master promotion.

**Comment by srajama1** [comment]
@kyungjoo-kim Can you take care of this please ?

**Comment by kyungjoo-kim** [comment]
@crtrott In your dot testing, why do you use the following views ? 

```
    typedef Kokkos::View<ScalarA*[2],                                                                                                                        
       typename std::conditional<                                                                                                                            
                std::is_same<typename ViewTypeA::array_layout,Kokkos::LayoutStride>::value,                                                                  
                Kokkos::LayoutRight, Kokkos::LayoutLeft>::type,Device> BaseTypeA;                                                                            
    BaseTypeA b_a("A",N);                                                                                                                                    
    ViewTypeA a = Kokkos::subview(b_a,Kokkos::ALL(),0);                                                                                                      
```
In this way, the view is strided and it never tests any tpls as tpls require unit stride according to your ETI design.

**Comment by mhoemmen** [comment]
@kyungjoo-kim `_DOT` in the BLAS takes strides (`INCX` and `INCY`).

**Comment by kyungjoo-kim** [comment]
@mhoemmen I know that the standard blas interface has the stride interface but kokkoskernels ETI system instanciates for layoutleft (using unification of layout for 1D view layout right and layout left goes into the instanciated code but the strided case uses kokkos native implementation).

**Comment by kyungjoo-kim** [comment]
@crtrott I confused view semantics and thought subview will return stride view. It is all oaky.

**Comment by ndellingwood** [comment]
Should be addressed by PR #314, thanks @kyungjoo-kim and @mhoemmen !
